infra/app/adminweb.bicep

```diff
@@ -0,0 +1,40 @@
+param name string
+param location string = resourceGroup().location
+param tags object = {}
+
+param allowedOrigins array = []
+param appServicePlanId string
+param appCommandLine string = 'python -m streamlit run Admin.py --server.port 8000 --server.address 0.0.0.0 --server.enableXsrfProtection false'
+param applicationInsightsName string = ''
+param keyVaultName string = ''
+param azureOpenAIName string = ''
+param azureCognitiveSearchName string = ''
+
+@secure()
+param appSettings object = {}
+param serviceName string = 'adminweb'
+
+module adminweb '../core/host/appservice.bicep' = {
+  name: '${name}-app-module'
+  params: {
+    name: name
+    location: location
+    tags: union(tags, { 'azd-service-name': serviceName })
+    allowedOrigins: allowedOrigins
+    appCommandLine: appCommandLine
+    applicationInsightsName: applicationInsightsName
+    appServicePlanId: appServicePlanId
+    appSettings: union(appSettings, {
+      AZURE_OPENAI_KEY: listKeys('Microsoft.CognitiveServices/accounts/${azureOpenAIName}', '2023-05-01').key1
+      AZURE_SEARCH_KEY: listAdminKeys('Microsoft.Search/searchServices/${azureCognitiveSearchName}', '2021-04-01-preview').primaryKey
```

Comment:
This only works if the openai and search are in the same resource group.  You will have to pass the keys into this from main.bicep or better yet, put the keys in keyvault and then just store the names here to be retrieved by the application.

Reply:
Hi, @jongio - We've tried a lot of things here on how to set `keys` to `appSettings`. Before making this change, I'd like to discuss best practices with you.

1.If we want to set `resources.keys` to `appSettings` in the `main.bicep` file, there are three methods we have tried so far:
- Create relevant resources directly in the main.bicep file, and then use `resources.listKeys().key1`. In this way, we cannot use templates in `core` to create related resources.
- Use `module` in `main.bicep` to reference related templates to create resources, but at this time, `resource.listKeys().key1` cannot be used because the resource is not created at the level of `main.bicep`; `listKeys(resourceId).key1` cannot be used either ( At present, the priority of the `listKeys` method is higher than that of module, unless we create the resource first).
- After we use the `module` to create the resource, use `existing` in `main.bicep` to call the relevant resources and then use `resource.listKeys().key1`. But this will also cause problems. The priority of `existing` is higher than `module`. We need to ensure that the resource is created in advance before we use `existing`.

Based on the results of our investigation so far, if you have any questions or if you have other best practices on how to set keys to appSettings in main.bicep, please let us know. Thank you very much!

2.About put the `keys` in `keyvault` and then just store the names here to be retrieved by the application:

- We also tried this in another azdevify [PR](https://github.com/Azure-Samples/azure-sql-db-session-recommender/pull/7/files/79acbab1a3fafca49c517befecc6188da1afb2c0#diff-ec8f897aa232a01946f69fd44fea3f1b350a0ae5fa1228d62bcb17bba705a239R45-R51). We refer to the approach in [todo template](https://github.com/Azure-Samples/todo-csharp-sql/blob/main/infra/core/database/cosmos/cosmos-account.bicep).  If we want to store the key of a resource, we need to reference `KeyVault` when creating the resource. Currently, the template for creating `openAi` in core does not have this function, which means that we cannot use the template in core when creating related resources.


Do you have any ideas on this? Our current idea is to add the function of storing `resource's keys` in the relevant templates in core: add a param `keyVaultName`, which is empty by default. If this value is empty, the resource `'Microsoft.KeyVault/vaults/secrets@2022-07-01'` will not be called.  If `keyVaultName` is passed in externally, the key of this resource can be stored in `KeyVault`. Do you think so?

Reply:
@jongio Any ideas about this issue?

Reply:
1. Add a parameter to the main.bicep that allows the user to either use RBAC or KEYS.  Call it AUTH_TYPE and default it to rbac.  And set @allowed to 'rbac' or 'keys'
2. If the auth type is rbac, then use roles.  If auth type is keys, then use keys.  

You can look up the keys with this listKeys method.  Example for storage.

```
listKeys(resourceId(subscription().subscriptionId, rgName, 'Microsoft.Storage/storageAccounts', storageAccountName), '2021-09-01').keys[0].value
```


It's like what you had before, but you also include the subscriptionId and RgName

Reply:
Currently we use `KeyVault` to store all keys and then obtain the keys from the application. What do you think of this? If you insist on using `rbac`, the way of creating various `service clients` in the application code needs to be changed accordingly. Then the gap between local deployment and deployment using `azd` will be slightly too large.

Reply:
I see what you mean.  Could you come up with a version of the code that shows what it would look like to instantiate the objects based on rbac or keys?  Just so I can see what the code would look like.  Doesn't have to be for them all.

Reply:
@jongio - Sorry for the delay. We have done some exploration and attempts on how to use `rbac`. Details as following:

Currently, a total of five `Keys` are used in our appSettings, which are: `AZURE_OPENAI_KEY`, `AZURE_SEARCH_KEY`, `AZURE_BLOB_ACCOUNT_KEY`, `AZURE_FORM_RECOGNIZER_KEY`, `AZURE_CONTENT_SAFETY_KEY`.

We divide it into three categories:

1. Cognitive Services (`AZURE_OPENAI_KEY`, `AZURE_FORM_RECOGNIZER_KEY`, `AZURE_CONTENT_SAFETY_KEY`):

- For this type of `Keys`, we can add rbac permissions to `openai`, and then use the following method to assign values to the `Keys`:
![image](https://github.com/jongio/chat-with-your-data-solution-accelerator/assets/81678720/67c1737f-7805-485a-ad73-e276a0924334)

2. AZURE_BLOB_ACCOUNT_KEY:

- We only need to add relevant permissions to `storage` and change the way to create `BlobServiceClient`:
![image](https://github.com/jongio/chat-with-your-data-solution-accelerator/assets/81678720/31b17301-1271-4f83-99ee-fc33cdf027f4)

3. AZURE_SEARCH_KEY:

- It needs to be used to create the request body in the [application](https://github.com/jongio/chat-with-your-data-solution-accelerator/blob/main/app.py#L79). We have not found a suitable replacement method yet: 
![image](https://github.com/jongio/chat-with-your-data-solution-accelerator/assets/81678720/a194f10b-3b65-4664-bed6-e98d825a72d0)If we want to change it, we can refer to [azure-search-openai-demo](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/app/backend/app.py), but the change will be large.

Reply:
Oh, I didn't know they aren't using the search SDK.  I'm asking them about that and will get back to you.